Add spancat_singlelabel pipeline for multiclass and non-overlapping span labelling tasks (#11365)
explosion/spaCy: website/docs/api/spancategorizer.mdx

A span categorizer consists of two parts: a [suggester function](#suggesters)
that proposes candidate spans, which may or may not overlap, and a labeler model
that predicts zero or more labels for each candidate. 

This component comes in two forms: `spancat` and `spancat_exclusive`.  When you
need to perform multi-label classification on your spans, use `spancat`.  The
`spancat` component uses a `Logistic` layer where the output class probabilities
are independent for each class.  However, if you need to predict exactly one true

> class for a span, then use `spancat_exclusive`. It uses a `Softmax` layer and treats

Reviewer: exactly one? Or "at most one"?

Author: I think "at most one" is more explicit. I'll update!

that proposes candidate spans, which may or may not overlap, and a labeler model
that predicts zero or more labels for each candidate.

> This component comes in two forms: `spancat` and `spancat_singlelabel`. When you
> need to perform multi-label classification on your spans, use `spancat`. The
> `spancat` component uses a `Logistic` layer where the output class probabilities
> are independent for each class. However, if you need to predict at most one true
> class for a span, then use `spancat_singlelabel` <Tag variant="new">3.5.1</Tag>.

Reviewer 2: This is a bit tricky, we don't usually use the new tag in running text. Something like this (and then formatted)?

```suggestion
This component comes in two forms: `spancat` and `spancat_singlelabel` (added in spaCy v3.5.1). When you
need to perform multi-label classification on your spans, use `spancat`. The
`spancat` component uses a `Logistic` layer where the output class probabilities
are independent for each class. However, if you need to predict at most one true
class for a span, then use `spancat_singlelabel`.
```

> treats the entities as a multi-class problem.

Reviewer 2: ```suggestion
treats the spans as a multi-class problem.
```

> nlp.add_pipe("spancat", config=config)
> ```

> #### Example (spancat_singlelabel)
>
> ```python
> from spacy.pipeline.spancat import DEFAULT_SPANCAT_SINGLELABEL_MODEL
> config = {
>     "threshold": 0.5,
>     "spans_key": "labeled_spans",

> >     "model": DEFAULT_EXCL_SPANCAT_MODEL,

Reviewer 2: Is this correct given the import above?

>     "suggester": {"@misc": "spacy.ngram_suggester.v1", "sizes": [1, 2, 3]},

> >     # Additional spancat_exclusive parameters
> >     "negative_weight": 0.8,
> >     "allow_overlap": True,
> > }
> > nlp.add_pipe("spancat_exclusive", config=config)

Reviewer 2: ```suggestion
>     # Additional spancat_singlelabel parameters
>     "negative_weight": 0.8,
>     "allow_overlap": True,
> }
> nlp.add_pipe("spancat_singlelabel", config=config)
```

| `threshold`    | Minimum probability to consider a prediction positive. Spans with a positive prediction will be saved on the Doc. Defaults to `0.5`. ~~float~~                                                                                                                                                          |
| `max_positive` | Maximum number of labels to consider positive per span. Defaults to `None`, indicating no limit. It is only available for the `spancat` component. ~~Optional[int]~~                                                                                                                                                                                      |
| `scorer`       | The scoring method. Defaults to [`Scorer.score_spans`](/api/scorer#score_spans) for `Doc.spans[spans_key]` with overlapping spans allowed. ~~Optional[Callable]~~                                                                                                                                       |

> | `negative_weight`    | Multiplier for the loss terms. It can be used to downweight the negative samples if there are too many. It is only available for the `spancat_exclusive` component. ~~float~~                                                  |

Reviewer: Add that it's `1.0` by default

> | `allow_overlap`        | If `True`, the data is assumed to contain overlapping spans. It is only available for the `spancat_exclusive` component. ~~bool~~ |

Reviewer: Needs `prettier` formatting

Reviewer: And add that it's `True` by default
